### PR TITLE
Bug fixing

### PR DESCRIPTION
**Pull request by FelixNeutatz**
The accuracy is now as good as in the original version of the extractor (even slightly better). There is probably still more space for performance improvements.

### CODE REVIEW

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 1]
```diff
@@ -1,17 +1,17 @@
 /**
  * Track the trackers
  * Copyright (C) 2015  Sebastian Schelter, Felix Neutatz
- *
+ * <p/>
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: remove

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
- *
+ * <p/>
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: remove

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 3]
```diff
@@ -1,17 +1,17 @@
...
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
- *
+ * <p/>
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: remove

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 4]
```diff
@@ -99,7 +99,8 @@
           ParserRunner.ParseResult parseResult = javascriptParser.parse(scriptContents);
           findUrlCandidates(parseResult.ast, javaScriptUrlCandidates);
         }
-      } catch (Exception e) {}
+      } catch (Exception e) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: we can leave the braces on the same line, we don't get anything from the extra line

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 5]
```diff
@@ -141,11 +142,10 @@ private void findUrlsInCode(List<String> candidateUrls) {
     candidateUrls.addAll(urlsInCode);
   }
 
-
   private Set<Resource> resourcesFromCandidates(List<String> candidateUrls) {
     Set<Resource> resources = Sets.newHashSet();
     for (String url : candidateUrls) {
-      if (URLHandler.couldBeUrl(url)) {
+      if (URLHandler.couldBeUrl(url.trim())) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: please trim the urls before putting them into candidateUrls

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
  * Track the trackers
  * Copyright (C) 2015  Sebastian Schelter, Felix Neutatz
- *
+ * <p/>
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
- *
+ * <p/>
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: remove

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 7]
```diff
@@ -1,17 +1,17 @@
 /**
  * Track the trackers
  * Copyright (C) 2015  Sebastian Schelter, Felix Neutatz
- *
+ * <p/>
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: remove

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 8]
```diff
@@ -1,17 +1,17 @@
...
  * the Free Software Foundation, either version 3 of the License, or
  * (at your option) any later version.
- *
+ * <p/>
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
- *
+ * <p/>
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: remove

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 9]
```diff
@@ -25,10 +25,10 @@
 
 class URLHandler {
 
-  private URLHandler() {}
+  private URLHandler() {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: leave the braces on the same line

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 10]
```diff
@@ -74,13 +80,30 @@ public static String createPrefixForInternalLinks(String sourceUrl) {
     return prefixForInternalLinks;
   }
 
+  public static String cleanURL(String url) {
+    String newUrl = new String(url);
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: remove, unnecessary, just change the next line to newUrl = url....

**Comment by sscdotopen** [comment]
the pull request looks good, please add a testcase that would fail with the current code and now works with the updated code. Please also run the MiniBenchmark and see if your changes influence the performance somehow.

**Comment by FelixNeutatz** [comment]
Mini Benchmark

My new code:
Runs: 1000 - execution time: 30880ms
ms per Run: 27
File size(in kB) / ms: 15
File size(in character) / ms: 16161
Script size(in character) / ms: 5690

My old code:
Runs: 1000 - execution time: 40247ms
ms per Run: 40
File size(in kB) / ms: 10
File size(in character) / ms: 11028
Script size(in character) / ms: 3883

**Comment by sscdotopen** [comment]
could you also compare it to the latest head?

**Comment by FelixNeutatz** [comment]
Current HEAD:
Runs: 1000 - execution time: 31556ms
ms per Run: 31
File size(in kB) / ms: 13
File size(in character) / ms: 14065
Script size(in character) / ms: 4952

**Comment by sscdotopen** [comment]
looks good. why did you change the version of jsoup?

**Comment by FelixNeutatz** [comment]
In some cases there were some strange parsing results, so I changed it to the newest version

**Comment by FelixNeutatz** [comment]
Performance still the same, but we can get a few more urls, and a few less messy urls

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 11]
```diff
@@ -74,13 +79,37 @@ public static String createPrefixForInternalLinks(String sourceUrl) {
...
+    return newUrl;
+  }
+
+  public static String clearDomain(String domain) throws MalformedURLException {
+    
+    domain = domain.replace("\\.", ".");
+       
+    String splits[] = domain.split("=");
+    
+    if (domain.contains("(") || domain.contains(")") || domain.contains("+") || domain.contains("\\")) {
```
Comment (FelixNeutatz) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: Shall we check here for more invalid characters:

> When choosing the name for your domain, always remember that: - you *can't* use stressed  vowels (such as à, é, ò, etc.); - you *can't* use symbols (such as ' + . , | ! " £ $ % & / ( ) = ? ^ * ç ° § ; : _ > ] [ @ ); - the name's length must range between 3 and 63 characters (excluding the extension); - the name can neither start nor end with the character "-", although the character "-" is allowed inside the name. So, to name your domain you can use any letter, numbers between 0 and 9, and the symbol "-". .
> 
> Length may vary, from 3 to 63 types.

**Comment by FelixNeutatz** [comment]
don't merge this yet, there are still some improvements

**Comment by FelixNeutatz** [comment]
I did a major rework. Now the parsing of the domains is done manually. This makes things way cleaner.

Current benchmark performance:
Runs: 1000 - execution time: 33771ms
ms per Run: 33
File size(in kB) / ms: 12
File size(in character) / ms: 13143
Script size(in character) / ms: 4627

Current result for the commoncrawl example:

2> (0,Google,836)
2> (0,Facebook,240)
2> (0,Twitter,129)
2> (0,AddThis,108)
2> (0,Yahoo,71)
2> (0,Amazon,29)
2> (0,StatCounter,25)
2> (0,PhotoBucket,13)
2> (0,CasaleMedia,8)

This seems also reasonable :)

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 12]
```diff
@@ -32,21 +32,13 @@
 
 import java.net.MalformedURLException;
 
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Set;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
+import java.util.*;
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: no star imports please

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 13]
```diff
@@ -69,7 +61,7 @@
 
     for (Element element : elements) {
       uri = element.attr("src").trim();
-
+
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: remove blank line

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 14]
```diff
@@ -74,31 +57,58 @@ public static String createPrefixForInternalLinks(String sourceUrl) {
...
+    boolean isDot = false;
+    int startDomain = 0;
+    for (int i = 0; i < len; i++) {
+      if (url.charAt(i) == '=') { //for javascript parsing src "=" domain -> to get a starting point
+        if (startDomain <= i) {
+          startDomain = i+1;
+        }
+      }
+      
+      if(isDot) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: space after if

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 15]
```diff
@@ -74,31 +57,58 @@ public static String createPrefixForInternalLinks(String sourceUrl) {
     return prefixForInternalLinks;
   }
 
+  public static String parseDomain(String url) {
+    
+    //parse domain from URL manually
+    String nUrl = "";
+    final int len = url.length();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: why is this final?

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 16]
```diff
@@ -74,31 +57,58 @@ public static String createPrefixForInternalLinks(String sourceUrl) {
...
+    boolean isDot = false;
+    int startDomain = 0;
+    for (int i = 0; i < len; i++) {
+      if (url.charAt(i) == '=') { //for javascript parsing src "=" domain -> to get a starting point
+        if (startDomain <= i) {
+          startDomain = i+1;
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: spaces around +

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java (hunk 14 above)
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: space after if

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 17]
```diff
@@ -74,31 +57,58 @@ public static String createPrefixForInternalLinks(String sourceUrl) {
...
+        }
+      }
+      
+      if(isDot) {
+        if (url.charAt(i) == '/') { // find google.com "/" to get the end point of domain
+          nUrl = url.substring(startDomain,i);
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: space after ,

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java [hunk 18]
```diff
@@ -74,31 +57,58 @@ public static String createPrefixForInternalLinks(String sourceUrl) {
...
+          return nUrl;        
+        }
+        if (url.charAt(i) == ':') {  // find google.com ":" 8080  to get the end point of domain
+          nUrl = url.substring(startDomain,i);
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLHandler.java]: space after ,

File: extraction/src/test/java/io/ssc/trackthetrackers/extraction/resources/URLHandlerTest.java [hunk 19]
```diff
@@ -20,12 +20,49 @@
 
 import org.junit.Test;
 
+import java.net.MalformedURLException;
+
 import static org.junit.Assert.assertTrue;
 
 public class URLHandlerTest {
```
Comment (sscdotopen) [inline extraction/src/test/java/io/ssc/trackthetrackers/extraction/resources/URLHandlerTest.java]: why do we catch Exceptions here, shouldn't the test fail?

**Comment by sscdotopen** [comment]
finally merged, sorry that it took so long. Lots of things on my plate :)

**Pull request merged**